pkg/certificate/transitions.go

```diff
@@ -0,0 +1,143 @@
+	d := &deferredCertificateStatusUpdate{
+		wait:                    wait,
+		certificateTypes:        certificateTypes,
+		mrcClient:               mrcClient,
+		targetMRCName:           targetMRCName,
+		targetCertificateStatus: targetCertStatus,
+	}
+
+	d.certificateStatusUpdates = map[CertificateType]func() error{
+		CertificateTypeSidecar: d.genUpdateFunc(func(mrc *v1alpha2.MeshRootCertificate, status v1alpha2.MeshRootCertificateComponentStatus) bool {
+			if mrc.Status.ComponentStatuses.Sidecar == status {
+				return false
+			}
+			mrc.Status.ComponentStatuses.Sidecar = status
+
+			return true
+		}),
+		CertificateTypeBootstrap: d.genUpdateFunc(func(mrc *v1alpha2.MeshRootCertificate, status v1alpha2.MeshRootCertificateComponentStatus) bool {
+			if mrc.Status.ComponentStatuses.Bootstrap == status {
+				return false
+			}
+			mrc.Status.ComponentStatuses.Bootstrap = status
+
+			return true
+		}),
+		CertificateTypeMutatingWebhook: d.genUpdateFunc(func(mrc *v1alpha2.MeshRootCertificate, status v1alpha2.MeshRootCertificateComponentStatus) bool {
+			// TODO: Change to mutatingWebhook specific field when MRC API changes are merged
+			if mrc.Status.ComponentStatuses.Webhooks == status {
+				return false
+			}
+			mrc.Status.ComponentStatuses.Webhooks = status
+
+			return true
+		}),
+		CertificateTypeValidatingWebhook: d.genUpdateFunc(func(mrc *v1alpha2.MeshRootCertificate, status v1alpha2.MeshRootCertificateComponentStatus) bool {
+			// TODO: Change to validatingWebhook specific field when MRC API changes are merged
+			if mrc.Status.ComponentStatuses.Webhooks == status {
+				return false
+			}
+			mrc.Status.ComponentStatuses.Webhooks = status
+
+			return true
+		}),
+		CertificateTypeXDSControlPlane: d.genUpdateFunc(func(mrc *v1alpha2.MeshRootCertificate, status v1alpha2.MeshRootCertificateComponentStatus) bool {
+			if mrc.Status.ComponentStatuses.XDSControlPlane == status {
+				return false
+			}
+			mrc.Status.ComponentStatuses.XDSControlPlane = status
+
+			return true
+		}),
+		CertificateTypeGateway: d.genUpdateFunc(func(mrc *v1alpha2.MeshRootCertificate, status v1alpha2.MeshRootCertificateComponentStatus) bool {
+			if mrc.Status.ComponentStatuses.Gateway == status {
+				return false
+			}
+			mrc.Status.ComponentStatuses.Gateway = status
+
+			return true
+		}),
+	}
+
+	return d
+}
+
+func (d *deferredCertificateStatusUpdate) genUpdateFunc(set setFunc) func() error {
+	return func() error {
+		mrc := d.mrcClient.GetMeshRootCertificate(d.targetMRCName)
+		if mrc == nil {
+			return fmt.Errorf("no MRC found with name %s in the osm control plane namespace", d.targetMRCName)
+		}
+
+		if !set(mrc, d.targetCertificateStatus) {
+			return nil
+		}
+
+		_, err := d.mrcClient.UpdateMeshRootCertificateStatus(mrc)
```

Comment:
this issues N update calls in consecutive order with no waits. That's going to trigger a lot of udpates. 

Is the intent that it does this with no waiting, or is the idea to perform a check between each one? Can you add a comment?

Reply:
Ah good point; we should queue up the calls before making a final update call. Good catch!